other/docker/misra/Makefile

```diff
@@ -0,0 +1,166 @@
+# See the following PDF for descriptions of each of the rules:
+# http://my.ldrasoftware.co.uk/repository/miscellaneous/Misra-c_2012_compliance.pdf
+
+# There should be no unused parameters in functions.
+#
+# Reason: callbacks often have unused parameters. Marking them explicitly isn't
+# very helpful. A better diagnostic should be able to identify functions never
+# used as callbacks and warn about unused parameters in those.
+SUPPRESSIONS = 2.7
+# The character sequences /* and // shall not be used within a comment.
+#
+# Reason: "//" appears in code examples and "http://" inside comments.
+SUPPRESSIONS += 3.1
+# Identifiers declared in the same scope and name space shall be distinct.
+# Identifier not unique within 31 characters.
+#
+# Reason: Compilers we use allow longer identifier names.
+SUPPRESSIONS += 5.2
+# Macro identifiers shall be distinct.
+# Identifier matches macro name in 31 chars.
+#
+# Reason: Compilers we use allow longer identifier names.
+SUPPRESSIONS += 5.4
+# The lowercase character 'l' shall not be used in a literal suffix.
+#
+# Reason: False positive. We don't use 'l', but this flags 'ulOutBufLen'.
+SUPPRESSIONS += 7.3
+# Operands shall not be of an inappropriate essential type.
+#
+# Reason: This diagnoses (1 << n) and wants us to use (1u << n). That's fair,
+# but this diagnostic is impossible to fix for ((1u << n) >> m).
+SUPPRESSIONS += 10.1
+# Both operands of an operator in which the usual arithmetic conversions are performed shall have the same essential type category.
+#
+# Reason: This warns about ((unsigned)n == 0) and other constant comparisons.
+SUPPRESSIONS += 10.4
+# The value of a composite expression shall not be cast to a different essential type category or a wider essential type.
+#
+# TODO(iphydf): investigate.
+SUPPRESSIONS += 10.8
+# A conversion should not be performed from pointer to void into pointer to object.
+#
+# Reason: this is needed for generic callbacks to make any sense.
```

Comment:
maybe we should keep our internal API free from casts to void? AFAICT this does not apply to objects passed in externally and then stored as `void *userdata`

Reply:
Lots of our internal APIs use internal callbacks. E.g. look for all the `(DHT *)` casts. It's everywhere. I'm not sure it's realistic to stop that, but maybe we can make a linter that says only objects named "object", "userdata", or "user_data" can be cast-from-void, to restrict the usage a bit?

Reply:
A linter for only `userdata` sounds like a good compromise. I'm not meaning to change this in the short term anyway.